src/esp/geo/VoxelGrid.cpp

```diff
@@ -198,129 +106,143 @@ Mn::Vector3 VoxelGrid::getGlobalCoords(const Mn::Vector3i& coords) {
-                                        bool (*func)(int)) {
+std::vector<Mn::Vector3i> VoxelGrid::fillVoxelSetFromIntGrid(
+    const std::string& intGridName,
+    int lb,
+    int ub) {
   assert(grids_.find(intGridName) != grids_.end());
+  std::vector<Mn::Vector3i> voxelSet = std::vector<Mn::Vector3i>();
   assert(grids_[intGridName].first == "int");
-  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
-                 m_voxelGridDimensions[2];
-  for (int i = 0; i < gridSize; i++) {
-    if (func(getVoxelByHash<int>(i, intGridName))) {
-      voxelSet.push_back(reverseHash(i));
+  auto intGrid = getGrid<int>(intGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (intGrid[i][j][k] >= lb && intGrid[i][j][k] <= ub) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
     }
   }
+  return voxelSet;
 }
 
-void VoxelGrid::fillVoxelSetFromFloatGrid(std::vector<Mn::Vector3i>& voxelSet,
-                                          const std::string& floatGridName,
-                                          bool (*func)(float)) {
+std::vector<Mn::Vector3i> VoxelGrid::fillVoxelSetFromFloatGrid(
+    const std::string& floatGridName,
+    float lb,
+    float ub) {
+  std::vector<Mn::Vector3i> voxelSet = std::vector<Mn::Vector3i>();
   assert(grids_.find(floatGridName) != grids_.end());
   assert(grids_[floatGridName].first == "float");
-  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
-                 m_voxelGridDimensions[2];
-  for (int i = 0; i < gridSize; i++) {
-    if (func(getVoxelByHash<float>(i, floatGridName))) {
-      voxelSet.push_back(reverseHash(i));
-    }
-  }
-}
-
-void VoxelGrid::fillVoxelSetFromVector3Grid(std::vector<Mn::Vector3i>& voxelSet,
-                                            const std::string& vector3GridName,
-                                            bool (*func)(Mn::Vector3)) {
-  assert(grids_.find(vector3GridName) != grids_.end());
-  assert(grids_[vector3GridName].first == "vector3");
-  int gridSize = m_voxelGridDimensions[0] * m_voxelGridDimensions[1] *
-                 m_voxelGridDimensions[2];
-  for (int i = 0; i < gridSize; i++) {
-    if (func(getVoxelByHash<Mn::Vector3>(i, vector3GridName))) {
-      voxelSet.push_back(reverseHash(i));
+  auto floatGrid = getGrid<float>(floatGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (floatGrid[i][j][k] >= lb && floatGrid[i][j][k] <= ub) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
     }
   }
+  return voxelSet;
 }
 
 // 6D SDF - labels each cell as interior (-inf), exterior (+inf), or boundary
 void VoxelGrid::generateInteriorExteriorVoxelGrid() {
```

Comment:
This is such a specific algorithm and specific application of VoxelGrid. From the class name, I would expect VoxelGrid to be a "dumb" container. Will this class grow and grow as we add more algorithms for operating on voxel grids? Can we instead make some kind of split here between the container and the algorithms? @aclegg3

Reply:
I think transfering all algorithms to the voxel wrapper would make sense, since I can't really imagine the case when someone would have a VoxelGrid without it's wrapper and need to use these algorithms.

I'm a bit hesitant about making the algorithms free functions, since it might make the python API a bit less intuitive (right now all you do is get the object's voxelWrapper and can call all functions using that). Although I'm open to an alternative design. Thoughts @aclegg3 @mosra ?

Reply:
@ScottCSteinhauser and I chatted about this a bit. Agreed that reducing bloat in the interface class would be ideal. We're thinking a voxel utilities module may be the best way to handle these sorts of applications/algorithms which can be implemented through the interface. Scott will break these out into a separate, incremental PR and we can focus on the interface first.